### PR TITLE
KEP-3619: Fine-grained SupplementalGroups control

### PR DESCRIPTION
**Pull request by everpeace**
For:

- https://github.com/kubernetes/enhancements/issues/3619

Depends on:

- k/k
  - [x] https://github.com/kubernetes/kubernetes/pull/117842
  - [x] https://github.com/kubernetes/kubernetes/pull/125470
- containerd (need to be [merged to main and released](https://github.com/kubernetes-sigs/cri-tools/blob/master/.github/workflows/containerd.yml#L29))
  - [x] https://github.com/containerd/containerd/pull/9737
  - [x] https://github.com/containerd/containerd/pull/10410
- CRI-O (need to be [merged to main](https://github.com/kubernetes-sigs/cri-tools/blob/master/.github/workflows/crio.yml#L16)):
  - [x] https://github.com/cri-o/cri-o/pull/8268
  - [x] https://github.com/cri-o/cri-o/pull/8386
- ~#1437 (v1.31 introduces `k8s.io/cri-client` isolated from kubelet)~

----

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Implements validation tests for KEP-3619: Fine-grained SupplementalGroups control:
- https://github.com/kubernetes/enhancements/issues/3619

, which skips itself when the container runtime does not support `SupplementalGroupsPolicy` detected via `StatusResponse.features.supplemental_groups_policy`.

The current SupplementalGroupsPolicy test status:

| CRI  | OS | Version | Test Stautus |
| ---  | --- | --- | --- |
| containerd | Linux | master | PASSED |
| containerd | Linux | 1.6 | SKIPPED |
| containerd | Linux | 1.7 | SKIPPED |
| containerd | Windows | * | SKIPPED |
| CRI-O | Linux | master | PASSED |
| CRI-O | Windows | * | SKIPPED |

#### Which issue(s) this PR fixes:

None

#### Does this PR introduce a user-facing change?

```release-note
Support Fine-grained SupplementalGroups control (KEP-3619)
```

### CODE REVIEW

**Review by everpeace — COMMENTED** [review]
@kwilczynski @mrunalp Could you kindly review this PR?? 🙇 Thanks, in advance.

File: pkg/validate/security_context_linux.go [hunk 1]
```diff
@@ -639,6 +589,151 @@ var _ = framework.KubeDescribe("Security Context", func() {
...
+					// - uid: 1000 (RunAsUser)
+					// - gid: 1000 (default group for uid=1000)
+					// - supplementary groups
+					//   - 1000: self
+					//   - 1234: SupplementalGroups
+					g.Expect(containerStatus.User).To(PointTo(MatchFields(IgnoreExtras, Fields{
+						"Linux": PointTo(MatchFields(IgnoreExtras, Fields{
+							"Uid": Equal(imagePredefinedGroupUID),
+							"Gid": Equal(imagePredefinedGroupUID),
+							// we can not assume the order of gids
```
Comment (kwilczynski) [inline pkg/validate/security_context_linux.go:715]: Would sorting the list make it a little more deterministic?

File: pkg/validate/security_context_linux.go [hunk 2]
```diff
@@ -639,6 +589,151 @@ var _ = framework.KubeDescribe("Security Context", func() {
...
+				containerConfig := &runtimeapi.ContainerConfig{
+					Metadata: framework.BuildContainerMetadata(containerName, framework.DefaultAttempt),
+					Image:    &runtimeapi.ImageSpec{Image: testImagePreDefinedGroup},
+					Command:  []string{"sh", "-c", "id -G; while :; do sleep 1; done"},
+					Linux: &runtimeapi.LinuxContainerConfig{
+						SecurityContext: &runtimeapi.LinuxContainerSecurityContext{
+							RunAsUser:          &runtimeapi.Int64Value{Value: imagePredefinedGroupUID},
+							SupplementalGroups: []int64{supplementalGroup},
+							// SupplementalGroupsPolicy_Merge is default(0)
+							// SupplementalGroupsPolicy: runtimeapi.SupplementalGroupsPolicy_Merge,
```
Comment (kwilczynski) [inline pkg/validate/security_context_linux.go:620]: Do you want to keep this comment here?

Comment (everpeace) [reply pkg/validate/security_context_linux.go]: fixed in https://github.com/kubernetes-sigs/cri-tools/pull/1438/commits/c45cf828ce27fc728c7e3342d87cd716d295be04

File: pkg/validate/security_context_linux.go (hunk 1 above)
Comment (everpeace) [reply pkg/validate/security_context_linux.go]: Sure, fixed in [c9678f4](https://github.com/kubernetes-sigs/cri-tools/pull/1438/commits/c9678f49ce66db0041b22d4ee59805df679ec169). Sorting removes gstruct usage, too👍.

**Review by everpeace — COMMENTED** [review]
@kwilczynski I addressed your comments. PTAL.

**Review by saschagrunert — APPROVED** [review]
Thanks!

/hold
for @kwilczynski final thoughts.

File: pkg/validate/security_context_linux.go [hunk 3]
```diff
@@ -639,6 +588,153 @@ var _ = framework.KubeDescribe("Security Context", func() {
...
+				podID, podConfig, podLogDir = createPodSandboxWithLogDirectory(rc)
+
+				By("create container for security context SupplementalGroups")
+				supplementalGroup := int64(1234)
+				containerName := "container-with-SupplementalGroupsPolicyMerge-" + framework.NewUUID()
+				logPath := containerName + ".log"
+				containerConfig := &runtimeapi.ContainerConfig{
+					Metadata: framework.BuildContainerMetadata(containerName, framework.DefaultAttempt),
+					Image:    &runtimeapi.ImageSpec{Image: testImagePreDefinedGroup},
+					Command:  []string{"sh", "-c", "id -G; while :; do sleep 1; done"},
```
Comment (kwilczynski) [inline pkg/validate/security_context_linux.go:680]: Would this work in place of the loop?

```suggestion
					Command:  []string{"sh", "-c", "id -G; sleep infinity"},
```

Comment (kwilczynski) [reply pkg/validate/security_context_linux.go:680]: I realised this is a code that existed prior... hence I apologise for asking about it. However, perhaps we can make things bit better. :smile:

File: pkg/validate/security_context_linux.go [hunk 4]
```diff
@@ -639,6 +588,153 @@ var _ = framework.KubeDescribe("Security Context", func() {
...
+
+				// $ id -G
+				// 1000 1234
+				expectedOutput := fmt.Sprintf("%d %d\n", imagePredefinedGroupUID, supplementalGroup)
+
+				By("verify groups for the first process of the container")
+				verifyLogContents(podConfig, logPath, expectedOutput, stdoutType)
+
+				By("verify groups for 'exec'-ed process of container")
+				command := []string{"id", "-G"}
```
[the remark this answers is not in the document]
Comment (everpeace) [reply pkg/validate/security_context_linux.go:731]: As far as I know [`GroupIds()`](https://pkg.go.dev/os/user#User.GroupIds) always parse `/etc/groups`. So, I think `GroupIds()` wouldn't work.

File: pkg/validate/security_context_linux.go (hunk 3 above)
Comment (everpeace) [reply pkg/validate/security_context_linux.go]: never mind. fixed in c9e3de68

File: pkg/validate/security_context_linux.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (kwilczynski) [reply pkg/validate/security_context_linux.go:731]: Got it. Thank you!

**Pull request merged**